code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm

```diff
@@ -95,6 +95,47 @@ GLOBAL_LIST_INIT(upgrade_categories, list("Buildings", "Defences", "Xenos"))//,
 	xeno_message("[buyer] has built a silo at [get_area(buildloc)]!", "xenoannounce", 5, buyer.hivenumber)
 	return ..()
 
+/datum/hive_upgrade/building/evotower
+	name = "Evolution Tower"
+	desc = "Constructs a tower that increases the rate of evolution point generation by 1.25 times per tower."
+	psypoint_cost = 300
+	icon = "evotower"
+	flags_upgrade = ABILITY_DISTRESS
+
+/datum/hive_upgrade/building/evotower/can_buy(mob/living/carbon/xenomorph/buyer, silent = TRUE)
+	. = ..()
+	if(!.)
+		return
+
+	var/turf/buildloc = get_step(buyer, SOUTHWEST)
+	if(!buildloc)
+		return FALSE
+
+	if(buildloc.density)
+		if(!silent)
+			to_chat(buyer, span_xenowarning("You cannot build in a dense location!"))
+		return FALSE
+
+/datum/hive_upgrade/building/evotower/on_buy(mob/living/carbon/xenomorph/buyer)
```

Comment:
honestly all these on_buy procs could be one proc on the parent proc
you just need a time to build, a building type, you already have the psy cost

Reply:
They have different build requirements

Reply:
not really

![image](https://user-images.githubusercontent.com/25491240/135612007-9f494593-8d4b-47a8-8996-19429ec32ad5.png)

![image](https://user-images.githubusercontent.com/25491240/135612071-1472f70a-d69b-49e5-b433-8fc108059d19.png)

The range check can be done on the child proc, but all the rest is identical

Reply:
also, are you sure you want to allow xeno to put up all towers close to each other?

Reply:
![image](https://user-images.githubusercontent.com/57223640/135666436-c4ad4a91-584c-4554-8a27-73b402e44140.png)

Reply:
fine by me